File: coding-test-practice/자바스크립트 알고리즘 문제풀이 입문(코딩테스트 대비)/섹션 9. 그래프와 탐색(DFS, BFS:너비우선탐색)/bfs.js
// BFS
// 상태 트리를 레벨탐색하고 최단거리를 구하는 방법이다.
// queue를 이용해서 구현한다.
// 출발지점에서 도착지점으로 가는 최단거리를 구할떄 BFS를 사용할 수 있다.
// 출발 지점에서 간선하나로 갈 수 있는 곳을 탐색하고 도착지점이 있는지 확인한다.
// 출발 지점에서 간선 두개로 갈 수 있는 곳을 탐색하고 도착지점이 있는지 확인한다.
// ... 각 레벨을 탐색하면서 최단거리를 구하면 된다.

// 아래 그림과 같은 이진트리를 넓이우선탐색해 보세요.
function solution() {
  let answer = ''
  let queue = []
  queue.push(1)

  while (queue.length) {
    let v = queue.shift()
    answer += v + ''
    for (let nv of [[v * 2, v * 2 + 1]]) {
      if (nv > 7) continue
      queue.push(nv)
    }
  }
  return answer
}

console.log(solution())
// 1 2 3 4 5 6 7

File: coding-test-practice/파이썬/암시적 그래프/BFS.py
from collections import deque

# 8 방향으로 이동 가능하다. 
def bfs(grid, r, c):
  row_len, col_len = len(grid), len(grid[0])
  visited = [[False] * col_len  for _ in range(row_len)]
  dr = [0, 1, 1, 1, 0, -1, -1, -1]
  dc = [1, 1, 0, -1, -1, -1, 0, 1]

  def isValid(r, c):
    return 0 <= r < row_len and 0 <= c < col_len and grid[r][c] == 1

  queue = deque()
  queue.append((r, c))

  visited[r][c] = True
  while queue:
    cur_r, cur_c = queue.popleft()

    for i in range(8):
      next_r = cur_r + dr[i]
      next_c = cur_c + dc[i]
      
      if isValid(next_r, next_c):
        if not visited[next_r][next_c]: # 여기 까지가 유효한 vertex 인가(길)를 증명하는 단계이다.
          queue.append((next_r, next_c))
          visited[next_r][next_c] = True

  print("방문 기록", visited)

grid = [[1, 1, 1, 1], [0, 1, 0, 1], [0, 1, 0, 1], [1, 0, 1, 1]]
bfs(grid, 0, 0)
